libs/client-api/src/v2/README.md

```diff
@@ -0,0 +1,929 @@
+
+2. **Concurrent Edits Phase**:
+    - While Client B is offline, Client A continues making changes
+    - Server state advances with Client A's updates
+    - State vector divergence occurs between clients
+
+3. **Reconnection Phase**:
+    - On reconnection, client sends a Manifest message with:
+        - Its current state vector [A:1, B:2]
+        - Last received message ID
+    - This triggers the `publish_pending_collabs` mechanism in the client
+
+4. **Reconciliation Phase**:
+    - Server compares client's state vector with its own
+    - Identifies missing updates on both sides
+    - Server discovers it needs to send [A:2] to client
+    - Server discovers it's missing [B:1, B:2] from client
+
+5. **Two-way Synchronization Phase**:
+    - Server sends missing updates to client
+    - Client sends its local changes to server
+    - Using `calculate_missing_update` function on the server
+
+6. **Propagation Phase**:
+    - Server broadcasts newly received updates to all other clients
+    - All clients converge to the same state [A:2, B:2]
+
+The offline editing capability is made possible by:
+
+- Local-first edits with CRDT data structures
+- State vector tracking to identify precisely what updates are missing
+- Two-way synchronization on reconnection
+- The `WorkspaceController` client implementation that manages connection state
+
+## Resilience Mechanisms
+
+The AppFlowy collaboration system implements several mechanisms to ensure reliability and consistency across various
+challenging scenarios.
+
+### Resource ID (Rid) Management
+
+```mermaid
+flowchart TD
+    subgraph "Rid Structure"
+        R[Resource ID]
+        R --> TS[timestamp: i64]
+        R --> SN[seq_no: u16]
+    end
+
+    subgraph "Comparison Logic"
+        CL[Rid Comparison]
+        CL --> T1{timestamp equal?}
+        T1 -->|Yes| S1[Compare seq_no]
+        T1 -->|No| T2[Compare timestamp]
+    end
+
+    subgraph "Client-Side Processing"
+        CM[Receive Update]
+        CM --> CR{Is Rid > current?}
+        CR -->|Yes| UR[Update last_message_id]
+        CR -->|No| KR[Keep current last_message_id]
+    end
+
+    style R fill: #f9f, stroke: #333
+    style CL fill: #ffe6cc, stroke: #333
+    style CM fill: #d1f0ff, stroke: #333
+```
+
+Resource IDs (Rid) provide a robust ordering mechanism for updates, ensuring consistency even when messages arrive out
+of order:
+
+```rust
+pub struct Rid {
+    pub timestamp: i64,  // UNIX timestamp when update was created
+    pub seq_no: u16,     // Sequence number for same-timestamp updates
+}
+
+impl Ord for Rid {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.timestamp.cmp(&other.timestamp) {
+            Ordering::Equal => self.seq_no.cmp(&other.seq_no),
+            ordering => ordering,
+        }
+    }
+}
+```
+
+#### Detailed Rid Implementation and Purpose
+
+The Rid mechanism is a critical component for maintaining causal consistency in the distributed CRDT system. It provides
```

Comment:
No, it's not. It's useful since all of the updates from all the clients come through the same Redis stream, so they are totally ordered. This is why we don't need to track causal consistency (via i.e. state vectors) when trying to read updates from the Redis Stream.